gax/src/main/java/com/google/api/gax/batching/v2/BatcherImpl.java

```diff
@@ -0,0 +1,189 @@
+    private RequestT prototype;
+
+    private Builder() {}
+
+    public Builder<ElementT, ElementResultT, RequestT, ResponseT> setBatchingDescriptor(
+        BatchingDescriptor<ElementT, ElementResultT, RequestT, ResponseT> batchingDescriptor) {
+      this.batchingDescriptor = batchingDescriptor;
+      return this;
+    }
+
+    public Builder<ElementT, ElementResultT, RequestT, ResponseT> setUnaryCallable(
+        UnaryCallable<RequestT, ResponseT> unaryCallable) {
+      this.unaryCallable = unaryCallable;
+      return this;
+    }
+
+    public Builder<ElementT, ElementResultT, RequestT, ResponseT> setPrototype(RequestT prototype) {
+      this.prototype = prototype;
+      return this;
+    }
+
+    public BatcherImpl<ElementT, ElementResultT, RequestT, ResponseT> build() {
+      return new BatcherImpl<>(this);
+    }
+  }
+
+  public static <EntryT, EntryResultT, RequestT, ResponseT>
+      Builder<EntryT, EntryResultT, RequestT, ResponseT> newBuilder() {
+    return new Builder<>();
+  }
+
+  /** {@inheritDoc} */
+  @Override
+  public ApiFuture<ElementResultT> add(final ElementT element) {
+    Preconditions.checkState(!isClosed, "Cannot perform batching on a closed connection");
+
+    if (batch == null) {
+      batch = new Batch(batchingDescriptor.newRequestBuilder(prototype));
+    }
+
+    SettableApiFuture<ElementResultT> result = SettableApiFuture.create();
+    batch.add(element, result);
+    return result;
+  }
+
+  /** {@inheritDoc} */
+  @Override
+  public void flush() throws InterruptedException {
+    sendBatch();
+    while (numOfRpcs.get() > 0) {
+      LockSupport.parkNanos(DEFAULT_FINISH_WAIT_NANOS);
+    }
+  }
+
+  /** sends accumulated elements asynchronously for batching. */
+  private void sendBatch() {
+    if (batch == null) {
+      return;
+    }
+    final Batch accumulatedBatch = batch;
+    batch = null;
+    numOfRpcs.incrementAndGet();
+
+    final ApiFuture<ResponseT> batchResponse =
+        callable.futureCall(accumulatedBatch.builder.build());
+
+    ApiFutures.addCallback(
+        batchResponse,
+        new ApiFutureCallback<ResponseT>() {
+          @Override
+          public void onSuccess(ResponseT response) {
+            batchingDescriptor.splitResponse(response, accumulatedBatch.results);
+            onCompletion();
+          }
+
+          @Override
+          public void onFailure(Throwable throwable) {
+            batchingDescriptor.splitException(throwable, accumulatedBatch.results);
+            onCompletion();
+          }
+        },
+        directExecutor());
+  }
+
+  private void onCompletion() {
+    numOfRpcs.decrementAndGet();
+  }
+
+  /** {@inheritDoc} */
+  @Override
+  public void close() throws InterruptedException {
+    isClosed = true;
+    flush();
+  }
+
+  /**
+   * This class represent one logical Batch. It accumulates all the elements and it's corresponding
+   * future element results for one batch.
+   */
+  class Batch {
```

Comment:
static?

Reply:
My intention was not to declare type params for inner class... but Have updated it to `private static class`.

Reply:
Please add a comment to make that clear